packages/flutter/lib/src/material/button_theme.dart

```diff
@@ -284,6 +330,382 @@ class ButtonThemeData extends Diagnosticable {
+  /// (continue to) gradually migrate to it.
+  final ColorScheme colorScheme;
+
+  /// The minimum size of a button's tap target.
+  ///
+  /// Null by default.
+  ///
+  /// See also:
+  ///
+  ///  * [getMaterialTargetTapSize], which is used by [RaisedButton],
+  /// [OutlineButton] and [FlatButton].
+  final MaterialTapTargetSize materialTapTargetSize;
+
+  /// The [button]'s overall brightness.
+  ///
+  /// Returns the button's [MaterialButton.colorBrightness] it it is non-null,
+  /// otherwise the color scheme's [ColorScheme.brightness] is returned.
+  Brightness getBrightness(MaterialButton button) {
+    return button.colorBrightness ?? colorScheme.brightness;
+  }
+
+  /// Defines the [button]'s base colors, and the defaults for the button's
+  /// minimum size, internal padding, and shape.
+  ///
+  /// Despite the name, this property is not the [TextTheme] whose
+  /// [TextTheme.button] is used as the button text's [TextStyle].
+  ButtonTextTheme getTextTheme(MaterialButton button) {
+    return button.textTheme ?? textTheme;
+  }
+
+  bool _isIconButton(MaterialButton button) {
+    return button is RaisedButtonWithIcon
+        || button is FlatButtonWithIcon
+        || button is OutlineButtonWithIcon;
+  }
+
+  Color _getDisabledColor(MaterialButton button) {
+    return getBrightness(button) == Brightness.dark
+      ? colorScheme.onSurface.withOpacity(0.30)  // default == Colors.white30
+      : colorScheme.onSurface.withOpacity(0.38); // default == Colors.black38;
+  }
+
+  /// The foreground color of the [button]'s text and icon when
+  /// [MaterialButton.onPressed] is null (when MaterialButton.enabled is false).
+  ///
+  /// Returns the button's [MaterialButton.disabledColor] if it is non-null.
+  /// Otherwise the color scheme's [ColorScheme.onSurface] color is returned
+  /// with its opacity set to 0.30 if [getBrightness] is dark, 0.38 otherwise.
+  Color getDisabledTextColor(MaterialButton button) {
+    if (button.disabledTextColor != null)
+      return button.disabledTextColor;
+    return _getDisabledColor(button);
+  }
+
+  /// The [button]'s background color when [MaterialButton.onPressed] is null
+  /// (when MaterialButton.enabled is false).
+  ///
+  /// Returns the button's [MaterialButton.disabledColor] if it is non-null.
+  /// Otherwise the color scheme's [ColorScheme.onSurface] color is returned
+  /// with its opacity set to 0.3 if [getBrightness] is dark, 0.38 otherwise.
+  Color getDisabledFillColor(MaterialButton button) {
+    if (button.disabledColor != null)
+      return button.disabledColor;
+    if (disabledColor != null)
+      return disabledColor;
+    return _getDisabledColor(button);
+  }
+
+  /// The button's background fill color or null for buttons that don't have
+  /// a background color.
+  ///
+  /// Returns [MaterialButton.color] if it is non-null and the button
+  /// is enabled.
+  ///
+  /// Otherwise, returns [MaterialButton.disabledColor] if it is non-null and
```

Comment:
For clarity, I would place the "If" first and drop the otherwise.

Reply:
I tried various permutations of this doc. It needs to convey the fact that the first non-null production is returned (I used "otherwise").  I agree that it's a little clunky as presented, and I wish the derivation of fillColor wasn't so complicated, but I haven't been able to improve it.